icu4c/source/i18n/unitsdata.cpp

```diff
@@ -0,0 +1,122 @@
+     */
+    explicit ConversionRateDataSink(MaybeStackVector<ConversionRateInfo> *out) : outVector(out) {}
+
+    /**
+     * Adds the conversion rate information found in value to the output vector.
+     *
+     * Each call to put() collects a ConversionRateInfo instance for the
+     * specified source unit identifier into the vector passed to the
+     * constructor, but only if an identical instance isn't already present.
+     *
+     * @param source The source unit identifier.
+     * @param value A resource containing conversion rate info (the base unit
+     * and factor, and possibly an offset).
+     * @param noFallback Ignored.
+     * @param status The standard ICU error code output parameter.
+     */
+    void put(const char *source, ResourceValue &value, UBool /*noFallback*/, UErrorCode &status) {
+        if (U_FAILURE(status)) return;
+        if (uprv_strcmp(source, "convertUnits") != 0) {
+            status = U_ILLEGAL_ARGUMENT_ERROR;
+            return;
+        }
+        ResourceTable conversionRateTable = value.getTable(status);
+        const char *srcUnit;
+        // We're reusing `value`, which seems to be a common pattern:
+        for (int32_t unit = 0; conversionRateTable.getKeyAndValue(unit, srcUnit, value); unit++) {
+            ResourceTable unitTable = value.getTable(status);
+            const char *key;
+            UnicodeString baseUnit = ICU_Utility::makeBogusString();
+            UnicodeString factor = ICU_Utility::makeBogusString();
+            UnicodeString offset = ICU_Utility::makeBogusString();
+            for (int32_t i = 0; unitTable.getKeyAndValue(i, key, value); i++) {
+                if (uprv_strcmp(key, "target") == 0) {
+                    baseUnit = value.getUnicodeString(status);
+                } else if (uprv_strcmp(key, "factor") == 0) {
+                    factor = value.getUnicodeString(status);
+                } else if (uprv_strcmp(key, "offset") == 0) {
+                    offset = value.getUnicodeString(status);
+                }
+            }
+            if (U_FAILURE(status)) return;
+            if (baseUnit.isBogus() || factor.isBogus()) {
+                // We could not find a usable conversion rate: bad resource.
+                status = U_MISSING_RESOURCE_ERROR;
+                return;
+            }
+
+            // We don't have this ConversionRateInfo yet: add it.
+            ConversionRateInfo *cr = outVector->emplaceBack();
+            if (!cr) {
+                status = U_MEMORY_ALLOCATION_ERROR;
+                return;
+            } else {
+                cr->sourceUnit.append(srcUnit, status);
+                cr->baseUnit.appendInvariantChars(baseUnit, status);
+                cr->factor.appendInvariantChars(factor, status);
+                if (!offset.isBogus()) cr->offset.appendInvariantChars(offset, status);
+            }
+        }
+        return;
+    }
+
+  private:
+    MaybeStackVector<ConversionRateInfo> *outVector;
```

Comment:
The sink is populating a MaybeStackVector it doesn't own. This seems to be a normal way of doing things - prior art review for a pointer or reference to the output data in a ResourceSink:

- ListFormatter::ListPatternsSink has a public char* read directly by its clients.
- RelDateFmtDataSink stores a pointer (`URelativeString *fDatesPtr;`),
- RelDateTimeFmtDataSink holds a reference passed to the constructor (`RelativeDateTimeCacheData &outputData;`).
- DecFmtSymDataSink holds a reference (`DecimalFormatSymbols& dfs;`)
- PluralTableSink holds a pointer (`UnicodeString *outArray;`)
- NumberRangeDataSink holds a reference (`NumberRangeData& fData;`), as does PluralRangesDataSink.
- ZNames::ZNamesLoader owns the loaded data and expects the client code to make a copy as needed.
- KeywordsSink (ucol_res.cpp) holds a pointer (`UList *values;`)

I went for a pointer, because stylistically I think it's clearer that there are "ownership concerns" (that one passes a pointer / an address is visible at the callsite).

(This new thread is a follow-on from https://github.com/sffc/icu/pull/32#discussion_r408212054 )